### PR TITLE
Security features

### PR DESCRIPTION
**Pull request by datdenkikniet**
This PR implements security for Data frames

I'm not entirely sure if I'm adhering to the design correctly, so I would highly appreciate feedback on that.

This PR is part of work performed for Widefind AB.

### CODE REVIEW

**Comment by hannobraun** [comment]
Thank you for the pull request, @datdenkikniet! I don't have the time to look into it right now, but I've made a note to check it out later, hopefully some time next week.

File: src/mac/frame/header.rs [hunk 1]
```diff
@@ -67,6 +68,34 @@ pub struct Header {
 
     /// Source Address
     pub source: Option<Address>,
+
+    /// Auxiliary security header, only Some if [`Header.security`] is true
+    pub auxiliary_security_header: Option<AuxiliarySecurityHeader>,
```
Comment (hannobraun) [inline src/mac/frame/header.rs:75]: Is this always going to be `Some`, if `security` is `true`. And `None`, if `security` is `false`? If so, can we just remove the `security` field?

File: src/mac/frame/header.rs [hunk 2]
```diff
@@ -8,7 +8,8 @@ use byte::{check_len, BytesExt, TryRead, TryWrite, LE};
 use hash32_derive::Hash32;
 
 use super::frame_control::*;
-pub use super::frame_control::{AddressMode, FrameType, FrameVersion, Security};
+pub use super::frame_control::{AddressMode, FrameType, FrameVersion};
+use super::security::*;
```
Comment (hannobraun) [inline src/mac/frame/header.rs]: I'm not a fan of these glob imports (except for limited preludes that mostly contain traits), as I think specifying the imported names explicitly makes the code more readable (especially if you don't have IDE support, like when reading code on GitHub).

Not a showstopper, just expressing my preference.

File: src/mac/frame/security/default.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    typenum::consts::{U1, U16},
+    GenericArray,
+};
+use cipher::{BlockCipher, NewBlockCipher};
+
+use super::{DeviceDescriptorLookup, KeyLookup};
+
+/// A struct that fullfills all of the type checks, but is not actually capable of
+/// performing any of the operations
+pub struct Unimplemented();
```
Comment (hannobraun) [inline src/mac/frame/security/default.rs]:
```suggestion
pub struct Unimplemented;
```

Not a big deal, just looks slightly cleaner.

**Review by hannobraun — COMMENTED** [review]
Hey @datdenkikniet, I did a quick review and this looks good. Although I have to admit I'm a bit clueless about IEEE 802.15.4 security (or IEEE 802.15.4 overall, to be honest :smile:), so I don't think I would spot many errors here. Still, if it works for you, I'm happy to merge it.

I've left a few comments. Just nitpicking, nothing serious.

Let me know when this is ready to get merged. I'll take a final look then.

**Comment by datdenkikniet** [comment]
Thank you for the review :) I'll continue working on this, fix the issues you've mentioned and and ping you once it is ready.

File: src/mac/frame/header.rs (hunk 1 above)
Comment (datdenkikniet) [reply src/mac/frame/header.rs:75]: Yes, and yes, that does seem like the more logical approach. Fixed

**Comment by datdenkikniet** [comment]
This PR is almost ready now. Most of the added structs and functionality has tests, and I've ran the software on a DW1000 chip to verify that it actually works, both with a wrapper for Chacha20 and AES. Both of these performed as expected.

I will add a more thorough how-to on actually using the security features, and then it should be good to go.

**Comment by datdenkikniet** [comment]
@hannobraun This PR is ready now :D I've tested this on a DW1000 (which worked fine) and added a usage example.

I assume that the version has to be updated to 0.6.0, but am holding off with changing it till I've received your feedback.

**Comment by hannobraun** [comment]
@datdenkikniet Great! I'll try to take a look later this week.

**Comment by hannobraun** [comment]
I haven't found the time to take a look this week. I'm hopeful I can do it next week. Sorry for the delay!

**Comment by hannobraun** [comment]
Finally found some time. Looks good to me! Thanks again, @datdenkikniet!

**Comment by datdenkikniet** [comment]
Thank you!

**Pull request merged**